workers/src/verifier.rs

```diff
@@ -0,0 +1,267 @@
+}
+
+// TODO: Generify in a future for other networks, like Solana.
+/// Verifies state roots for the Decentralized Verification Network (DVN).
+///
+/// This verifier implements the V1 verification algorithm which:
+/// 1. Fetches the state root from the aggregator
+/// 2. Retrieves the block state root using the chain RPC API
+/// 3. Compares the two state roots to determine message validity
+pub struct NFFLVerifier {
+    http_client: Client,
+    eth_l2_provider: ReqwestProvider<Ethereum>,
+    aggregator_http_address: String,
+    network_id: String,
+}
+
+impl NFFLVerifier {
+    pub(crate) async fn new(agg_url: &str, eth_l2_url: &str, network_id: u64) -> eyre::Result<NFFLVerifier> {
+        let client = ClientBuilder::new().build()?;
+        let url: Url = eth_l2_url.parse()?;
+        let provider = ProviderBuilder::new().on_http(url);
+        let agg_http_addr = format!("{}/state-root", agg_url);
+
+        Ok(NFFLVerifier {
+            eth_l2_provider: provider,
+            http_client: client,
+            aggregator_http_address: agg_http_addr,
+            network_id: network_id.to_string(),
+        })
+    }
+
+    pub async fn new_from_config(cfg: &DVNConfig) -> eyre::Result<NFFLVerifier> {
+        Self::new(&cfg.aggregator_url, &cfg.http_rpc_url, cfg.network_eid).await
+    }
+
+    /// Verifies the state root of a block. In case any request future
+    /// is interrupted, or finishes unsuccessfully, returns Ok(false).
+    pub async fn verify(&self, block_height: u64) -> eyre::Result<bool> {
+        const TIMEOUT: Duration = Duration::from_secs(10);
+        match tokio::try_join!(
+            tokio::time::timeout(TIMEOUT, self.get_aggregator_root_state(block_height)),
+            tokio::time::timeout(TIMEOUT, self.get_block_state_root(block_height)),
+        ) {
+            Ok((Ok(agg_response), Ok(block_state_root))) => {
+                let state_root_slice: &[u8] = agg_response.message.state_root.as_slice();
+                let aggregator_state_root: B256 = B256::from_slice(state_root_slice);
+                if agg_response.message.block_height != block_height {
+                    return Ok(false);
+                }
+                Ok(block_state_root.eq(&aggregator_state_root))
+            }
+            Err(e) => {
+                error!("Error while verifying state root: {:?}", e);
+                Ok(false)
+            }
+            _ => Ok(false),
+        }
+    }
+
+    /// Fetches the root state from the NFFL aggregator via HTTP.
+    pub(crate) async fn get_aggregator_root_state(&self, block_height: u64) -> eyre::Result<ResponseWrapper> {
+        let params = [
+            ("rollupId", self.network_id.clone()),
+            ("blockHeight", block_height.to_string()),
+        ];
+        let url = Url::parse_with_params(&self.aggregator_http_address, &params)?;
+        let response = self.http_client.get(url).send().await?;
+        let result = response.json::<ResponseWrapper>().await?;
```

Comment:
Why do we need a wrapper here, why can't we cast it to a message ?

Reply:
Because the JSON object we get wraps `Message`.